src/GraphQL/Instrumentation/FieldResolverBuilderExtensions.cs

```diff
@@ -8,42 +8,84 @@ namespace GraphQL.Instrumentation
+                    var instance = CheckSchemaNotNull(schema) is IServiceProvider provider
+                        ? CheckNotNull((IFieldMiddleware)provider.GetService(middleware), middleware)
+                        : throw NotSupported(schema);
 
-            return builder.Use(next =>
+                    return context => instance.Resolve(context, next);
+                });
+            }
+            // otherwise, we first find this method on the type and then call it through reflection
+            else
             {
-                var instance = Activator.CreateInstance(middleware);
+                var methods = middleware.GetMethods(BindingFlags.Instance | BindingFlags.Public);
+                var invokeMethods = methods.Where(m => string.Equals(m.Name, INVOKE_METHOD_NAME, StringComparison.Ordinal)).ToArray();
+                if (invokeMethods.Length > 1)
+                {
+                    throw new InvalidOperationException($"There should be only a single method named {INVOKE_METHOD_NAME}. Middleware actually has {invokeMethods.Length} methods.");
+                }
 
-                return context => (Task<object>)methodInfo.Invoke(instance, new object[] { context, next });
-            });
+                if (invokeMethods.Length == 0)
+                {
+                    throw new InvalidOperationException($"Could not find a method named {INVOKE_METHOD_NAME}. Middleware must have a public instance method named {INVOKE_METHOD_NAME}.");
+                }
+
+                var methodInfo = invokeMethods[0];
+                if (!typeof(Task<object>).IsAssignableFrom(methodInfo.ReturnType))
+                {
+                    throw new InvalidOperationException($"The {INVOKE_METHOD_NAME} method should return a Task<object>.");
+                }
+
+                var parameters = methodInfo.GetParameters();
+                if (parameters.Length != 2 || parameters[0].ParameterType != typeof(IResolveFieldContext) || parameters[1].ParameterType != typeof(FieldMiddlewareDelegate))
+                {
+                    throw new InvalidOperationException($"The {INVOKE_METHOD_NAME} method of middleware should take a parameter of type {nameof(IResolveFieldContext)} as the first parameter and a parameter of type {nameof(FieldMiddlewareDelegate)} as the second parameter.");
+                }
+
+                return builder.Use((schema, next) =>
+                {
+                    // Not an ideal solution, but at least it allows to work with custom schemas which are not inherited from Schema type
+                    object instance = CheckSchemaNotNull(schema) is IServiceProvider provider
+                        ? CheckNotNull(provider.GetService(middleware), middleware)
+                        : throw NotSupported(schema);
+
+                    return context => (Task<object>)methodInfo.Invoke(instance, new object[] { context, next });
```

Comment:
```suggestion
                    var resolver = (Func<IResolveFieldContext, FieldMiddlewareDelegate, Task<object>>)methodInfo.CreateDelegate(typeof(Func<IResolveFieldContext, FieldMiddlewareDelegate, Task<object>>), instance);
                    return context => resolver(context, next);
```
It is possible, for a minute gain, to compile the resolver with an additional line of code.  I did test it to be sure it works.

Reply:
Doesn't really matter to me.  People should use `IFieldMiddleware` !

Reply:
Right. If one want to optimize, then this is the best option (and easier).